src/tacacs/bash_tacplus/bash_tacplus.c

```diff
@@ -0,0 +1,483 @@
+    http://www.c-faq.com/varargs/handoff.html
+*/
+#define GENERATE_LOG_FROM_VA(logBufferName)                 \
+    char logBufferName[512];                                \
+    va_list args;                                           \
+    va_start(args, format);                                 \
+    vsnprintf(logBufferName, sizeof(logBufferName), format, args);  \
+    va_end(args);
+
+/* Config file path */
+const char *tacacs_config_file = "/etc/tacplus_nss.conf";
+
+/* Unknown user name */
+const char *unknown_username = "UNKNOWN";
+
+
+/* Config file attribute */
+struct stat config_file_attr;
+
+/* Tacacs server config data */
+typedef struct {
+    struct addrinfo *address;
+    const char *key;
+} tacacs_server_t;
+
+/* Tacacs control flag */
+int tacacs_ctrl;
+
+/*
+ * Output error message.
+ */
+void output_error(const char *format, ...)
+{
+    GENERATE_LOG_FROM_VA(logBuffer);
+
+    if (tacacs_ctrl & PAM_TAC_DEBUG) {
+        fprintf(stderr, TACACS_LOG_FORMAT, logBuffer);
+    }
+
+    syslog(LOG_ERR, TACACS_LOG_FORMAT, logBuffer);
+}
+
+/*
+ * Output debug message.
+ */
+void output_debug(const char *format, ...)
+{
+    if ((tacacs_ctrl & PAM_TAC_DEBUG) == 0) {
+        return;
+    }
+
+    GENERATE_LOG_FROM_VA(logBuffer);
+    fprintf(stderr, TACACS_LOG_FORMAT, logBuffer);
+    syslog(LOG_DEBUG, TACACS_LOG_FORMAT, logBuffer);
+}
+
+
+/*
+ * Send authorization message.
+ * This method based on send_auth_msg in https://github.com/daveolson53/tacplus-auth/blob/master/tacplus-auth.c
+ */
+int send_authorization_message(
+    int tac_fd,
+    const char *user,
+    const char *tty,
+    const char *host,
+    uint16_t taskid,
+    const char *cmd,
+    char **args,
+    int argc)
+{
+    char buf[128];
+    struct tac_attrib *attr;
+    int retval;
+    struct areply re;
+    int i;
+
+    attr=(struct tac_attrib *)xcalloc(1, sizeof(struct tac_attrib));
+
+    snprintf(buf, sizeof buf, "%hu", taskid);
+    tac_add_attrib(&attr, "task_id", buf);
+    tac_add_attrib(&attr, "protocol", "ssh");
+    tac_add_attrib(&attr, "service", "shell");
+
+    tac_add_attrib(&attr, "cmd", (char*)cmd);
+
+    for(i=1; i<argc; i++) {
+        // TACACS protocol allow max 255 bytes per argument. 'cmd-arg' will take 7 bytes.
+        char tbuf[248];
+        const char *arg;
+        if(strlen(args[i]) >= sizeof(tbuf)) {
+            snprintf(tbuf, sizeof tbuf, "%s", args[i]);
+            arg = tbuf;
+        }
+        else {
+            arg = args[i];
+        }
+
+        tac_add_attrib(&attr, "cmd-arg", (char *)arg);
+    }
+
+    re.msg = NULL;
+    output_debug("send authorizatiom message with user: %s, tty: %s, host: %s\n", user, tty, host);
+    retval = tac_author_send(tac_fd, (char *)user, (char *)tty, (char *)host, attr);
+    output_debug("authorization result: %d\n", retval);
+
+    if(retval < 0) {
+            output_error("send of authorization message failed: %s\n", strerror(errno));
```

Comment:
>             [](http://example.com/codeflow?start=0&length=12)

wrong indentation. If upstream code, keep it.

Reply:
Fixed.